pkg/agent/watchers.go

```diff
@@ -0,0 +1,81 @@
+package agent
+
+import (
+	"context"
+	"sync"
+
+	"github.com/datawire/k8sapi/pkg/k8sapi"
+	"github.com/emissary-ingress/emissary/v3/pkg/kates"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+)
+
+type Watchers struct {
+	cond            *sync.Cond
+	mapsWatcher     *k8sapi.Watcher[*kates.ConfigMap]
+	deployWatcher   *k8sapi.Watcher[*kates.Deployment]
+	podWatcher      *k8sapi.Watcher[*kates.Pod]
+	endpointWatcher *k8sapi.Watcher[*kates.Endpoints]
+}
+
+func NewWatchers(client rest.Interface) *Watchers {
+	// This function compares the recieved object with the cached object
+	// and decides whether an update should be pushed
+	compareFunc := func(o1, o2 runtime.Object) bool {
+		// TODO impl
+		return true
+	}
+
+	cond := &sync.Cond{
+		L: &sync.RWMutex{},
+	}
+
+	watchedNs := ""
+
+	return &Watchers{
+		mapsWatcher:     k8sapi.NewWatcher("configmap", watchedNs, client, &kates.ConfigMap{}, cond, compareFunc),
+		deployWatcher:   k8sapi.NewWatcher("deploy", watchedNs, client, &kates.Deployment{}, cond, compareFunc),
+		podWatcher:      k8sapi.NewWatcher("pod", watchedNs, client, &kates.Pod{}, cond, compareFunc),
+		endpointWatcher: k8sapi.NewWatcher("endpoints", watchedNs, client, &kates.Endpoints{}, cond, compareFunc),
```

Comment:
In the telepresence repo, we're using `k8s.io/api` modules with `k8sapi`; any reason we're not using them here?

Reply:
yeah... kates owns the snapshot struct, and therefore all the snapshot types. jose and i decided to leave it there for now, which means we have to produce objects of type kates.obj to put them in the snapshot